### PR TITLE
Add initial support to 'assets fetch' command with a simple parser [v2.1]

### PR DESCRIPTION
**Pull request by willianrampazzo**
Implements 'assets fetch' command that uses a parser on instrumented test source to find 'fetch_asset' calls composed of simple strings as parameters, or at least one level of a variable in the same context
with a string assignment, and fetch those assets without running the test.

Signed-off-by: Willian Rampazzo <willianr@redhat.com>

---

Changes from v1 (#3333)

- Executing asset.Asset instead of an avocado test.
- Handling one level of variable string assignment.
- Fix lots of pylint messages.
- Changed the way the call is built to fit asset.Asset call.

ps. fixed minor issue from v2. Still missing functional/unit tests as a v3 is expected.

### CODE REVIEW

**Comment by willianrampazzo** [comment]
[Implement fetch-assets command line](https://trello.com/c/T3SC1sZs/1521-implement-fetch-assets-command-line)

File: avocado/plugins/assets.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+"""
+Assets subcommand
+"""
+
+import ast
+import os
+
+from avocado.core import data_dir
+from avocado.core import safeloader
+
```
Comment (clebergnu) [inline avocado/plugins/assets.py]: Nitpick: the predominant style is to not have spaces between blocks importing from the same "namespace", that is, among all `avocado.*` modules.

File: avocado/plugins/assets.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+from avocado.core.output import LOG_UI
+from avocado.core.plugin_interfaces import CLICmd
+
+from avocado.utils import asset
+from avocado.utils import data_structures
+
+
+class FetchAssetHandler(ast.NodeVisitor):  # pylint: disable=R0902
+    """
+    Handles the parse of instrumented tests for `fetch_asset` statements
```
Comment (clebergnu) [inline avocado/plugins/assets.py]: I believe this should read "Handles the parsing ..." instead.

File: avocado/plugins/assets.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        fetch_subcommand_parser = subcommands.add_parser(
+            'fetch',
+            help='Fetch assets from test source or config file if it\'s not'
+            ' already in the cache')
+        fetch_subcommand_parser.add_argument('references', nargs='+',
+                                             metavar='TEST_REFERENCE',
+                                             help='List of test references')
+    # configure()
+
+    def run(self, config):
```
Comment (clebergnu) [inline avocado/plugins/assets.py]: It's important for this command to return a meaningful exit code to its callers.  I see this command being used before the `avocado run` invocation, in which a failure to fetch the assets would not attempt to run the tests (for minimum changes of false negatives due to network issues, download timeouts, etc).

Take a look at `avocado.core.exit_codes` for some documented exit statuses.

File: avocado/plugins/assets.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    Implements the avocado 'assets' subcommand
+    """
+    name = 'assets'
+    description = 'Manage assets'
+
+    def configure(self, parser):
+        """
+        Add the subparser for the assets action.
+
+        :param parser: Main test runner parser.
```
Comment (clebergnu) [inline avocado/plugins/assets.py]: I understand this docstring is in many other plugins, but it's actually a bit misleading.  The parser is the `avocado` command line application parser, which is a lot more than a "test runner", but the main entry point into most of Avocado functionality.  I'd go with:

```
:param parser: The Avocado command line application parser
:type parser: :class:`avocado.core.parser.ArgumentParser`
```

See https://github.com/avocado-framework/avocado/pull/3339 for the resolution on the other occurrences.

File: avocado/plugins/assets.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+from avocado.core import safeloader
+
+from avocado.core.output import LOG_UI
+from avocado.core.plugin_interfaces import CLICmd
+
+from avocado.utils import asset
+from avocado.utils import data_structures
+
+
+class FetchAssetHandler(ast.NodeVisitor):  # pylint: disable=R0902
```
Comment (clebergnu) [inline avocado/plugins/assets.py]: This class is a node visitor, but you're mixing it up with, what I consider to be, unrelated code that actually fetches the asset.  IMO, the two deserve to be separated, and `fetch_assets()` may be found to be enough to be a function.

This is a design decision, so make your best judgement about this comment.

File: avocado/plugins/assets.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        # discards disabled tests
+        self.tests = safeloader.find_avocado_tests(self.file_name)[0]
+
+        # create Abstract Syntax Tree from test source file
+        with open(self.file_name) as source_file:
+            self.tree = ast.parse(source_file.read(), self.file_name)
+
+        # build list of keyword arguments from calls that match pattern
+        self.visit(self.tree)
+    # __init__()
```
Comment (clebergnu) [inline avocado/plugins/assets.py]: The closing of methods/classes with comments is not present anywhere else in Avocado.  It's not wrong, but it's a good idea to conform to the project style as much as possible.  In short, I'd drop these.

File: avocado/plugins/assets.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+                # look for assignments at method
+                if arg.id in self.asgmts[cur_klass][cur_method]:
+                    args.append(self.asgmts[cur_klass][cur_method][arg.id])
+                # right now we support just one level of variable as argument,
+                # with a pure string assignment, in the same context,
+                # just like `name = 'file.zip'`
+                else:
+                    return None
+
+        # start building the keywords dictionary for the call
```
Comment (clebergnu) [inline avocado/plugins/assets.py]: You may want to explain that "the call" is `avocado.utils.asset.Asset()` (right?), which, is actually a class constructor.
